### PR TITLE
generalize Lambda_Acquire

### LINKED ISSUE #44: dm-support : writing to epics registers
@prjemian I am reviewing the DM commands to run to generate the metadata. It seems like we are missing the spec macro equivalent `write_CCD_MetaData_Registers` which fills in some of the registers.

***one step missing here before the 2 below***

    dm_pars = DataManagementMetadata(name="dm_pars")
    dm_workflow = APS_DM_8IDI.DM_Workflow(dm_pars)

### LINKED ISSUE #53: generalize for more than one area detector
We've been developing a plan for acquisition and data management workflow with the Lambda detector.  We want to execute that plan with more than one detector.  Some of the setup of device staging is specific to the Lambda detector.  That needs to be detector-specific code, under the Lambda Device.  Best to add a device-specific method plan here: https://github.com/aps-8id-trr/ipython-8idiuser/blob/b02826dbdabdf0752a3034fb9c73c28cacb8f941/profile_bluesky/startup/25-areadetector.py#L101-L118

### LINKED ISSUE #57: alphabetize attributes in DataManagementMetadata
Current list in order as defined in SPEC config file, that order is no longer relevant.  Sort alphabetically.

### LINKED ISSUE #62: resolve TODO & FIXME items in spec_support/APS_DM_8IDI.py
See items in the source code file marked `TODO:` or `FIXME:` (per #54)

### LINKED ISSUE #63: resolve TODO & FIXME items in 35-DM-support.py
See items in the source code file marked `TODO:` or `FIXME:` (per #54)

### PR DESCRIPTION
**Pull request by prjemian**
fixes #53, fixes #44, fixes #62, fixes #63, fixes #57

### CODE REVIEW

**Comment by prjemian** [comment]
@decarlof : You might be interested in this change.  Here, we localize setup steps for a specific area detector since other area detectors might have different setup steps.

**Comment by prjemian** [comment]
@qzhang234 , @sureshnaps : Look this over to see that you understand what was done.  Also, with the code in `Lambda_Acquire()` generalized, we might also generalize the name `Lambda_Acquire` as part of this PR.

**Comment by prjemian** [comment]
Once we pull this, we can add in the parts that set the string registers with the plan's values, such as `file_path`, `file_name`, ...

File: user/Lambda.py [hunk 1]
```diff
@@ -67,12 +55,30 @@ def Lambda_Acquire(acquire_time=0.1, acquire_period=0.11,num_images=100,file_nam
     @bpp.stage_decorator([scaler1])
     @bpp.monitor_during_decorator(monitored_things)
     def inner():
+        # write metadata to the dm_pars (various registers)
+        yield from bps.mv(
+            dm_pars.root_folder, file_path,
+            dm_pars.parent_folder, os.path.dirname(file_path),
+            dm_pars.data_folder, file_name,
+            dm_pars.datafilename, areadet.get_plugin_file_name(),
+            # TODO: what else?
```
Comment (qzhang234) [inline user/Lambda.py]: start_time, end_time to 8idi:StrReg6 and 7, after the acq, ring_i_beg and _end to 8idi:Reg121 and 122
registers
registers1

File: user/Lambda.py [hunk 2]
```diff
@@ -17,26 +14,17 @@
...
 """
 
-def Lambda_Acquire(acquire_time=0.1, acquire_period=0.11,num_images=100,file_name="A005"):
-    adlambda.cam.stage_sigs["num_images"] = num_images
-    adlambda.cam.stage_sigs["acquire_time"] = acquire_time
-    adlambda.cam.stage_sigs["acquire_period"] = acquire_period
-    adlambda.immout.stage_sigs["enable"] = 1
-    adlambda.immout.stage_sigs["blocking_callbacks"] = "Yes"
-    adlambda.immout.stage_sigs["parent.cam.array_callbacks"] = 1
+def Lambda_Acquire(areadet, acquire_time=0.1, acquire_period=0.11, num_images=100, file_name="A001"):
```
Comment (qzhang234) [inline user/Lambda.py]: change name to AD_Acquire

File: user/Lambda.py [hunk 3]
```diff
@@ -67,12 +55,30 @@ def Lambda_Acquire(acquire_time=0.1, acquire_period=0.11,num_images=100,file_nam
...
+
         md = {
             "file_name": file_name,
             "file_path": file_path
         }
-        yield from bps.mv(scaler1.count,"Count")
-        yield from bp.count([adlambda], md=md)
+        yield from bps.mv(scaler1.count, "Count")
+        yield from bp.count([areadet], md=md)
+
```
Comment (qzhang234) [inline user/Lambda.py]: set the post acq metadata

File: user/Lambda.py [hunk 4]
```diff
@@ -67,12 +55,30 @@ def Lambda_Acquire(acquire_time=0.1, acquire_period=0.11,num_images=100,file_nam
...
+        # FIXME: need correct filename here: hdf_with_fullpath
+        yield from dm_workflow.create_hdf5_file(hdf_with_fullpath, as_bluesky_plan=True)
+        
+        # kickoff the DM workflow
+        # TODO: which one?
+        # TODO: should run in thread, does it?
+        #yield from dm_workflow.DataTransfer(hdf_with_fullpath)
```
Comment (qzhang234) [inline user/Lambda.py]: in spec, we have a flag called SUBMIT_XPCS_JOB, if 1, then dm_analysis is called, if 0, then dm_transfer is called

File: user/Lambda.py [hunk 5]
```diff
@@ -67,12 +55,30 @@ def Lambda_Acquire(acquire_time=0.1, acquire_period=0.11,num_images=100,file_nam
...
             "file_name": file_name,
             "file_path": file_path
         }
-        yield from bps.mv(scaler1.count,"Count")
-        yield from bp.count([adlambda], md=md)
+        yield from bps.mv(scaler1.count, "Count")
+        yield from bp.count([areadet], md=md)
+
+        # FIXME: need correct filename here: hdf_with_fullpath
+        yield from dm_workflow.create_hdf5_file(hdf_with_fullpath, as_bluesky_plan=True)
```
Comment (qzhang234) [inline user/Lambda.py]: hdf file name:
file_name_%04d(starting_frame which is always 1)-%04d(num_frames).hdf

hdf file path:
file_path_network = file_path (replace /data with /home/8-id-i)
file_path_network + data_folder

File: user/Lambda.py [hunk 6]
```diff
@@ -67,12 +55,30 @@ def Lambda_Acquire(acquire_time=0.1, acquire_period=0.11,num_images=100,file_nam
...
+        
+        # kickoff the DM workflow
+        # TODO: which one?
+        # TODO: should run in thread, does it?
```
Comment (qzhang234) [inline user/Lambda.py]: these dm routines return information that should be logged

    id=1f131fd4-76e2-4e0f-b97f-d0acd809242b owner=8idiuser status=pending startTime=1567697398.43 startTimestamp=2019/09/05 10:29:58 CDT

File: user/Lambda.py (hunk 5 above)
Comment (qzhang234) [reply user/Lambda.py]: check if the file exists and create alternate name if file exists

File: user/Lambda.py (hunk 2 above)
Comment (prjemian) [reply user/Lambda.py]: done

File: user/Lambda.py (hunk 4 above)
Comment (prjemian) [reply user/Lambda.py]: will use a boolean, default=True

Comment (prjemian) [reply user/Lambda.py]: done

File: user/Lambda.py (hunk 5 above)
Comment (prjemian) [reply user/Lambda.py]: implemented in internal function

File: user/Lambda.py (hunk 6 above)
Comment (prjemian) [reply user/Lambda.py]: @sureshnaps Where should this be logged?  Along with the other logged output already generated?

Comment (prjemian) [reply user/Lambda.py]: I put it in the logging for now.  We can change this.

File: user/Lambda.py (hunk 3 above)
Comment (prjemian) [reply user/Lambda.py]: examine TODO items in `update_metadata_postscan()`

File: user/Lambda.py (hunk 1 above)
Comment (prjemian) [reply user/Lambda.py]: done but needs more discussion

File: user/Lambda.py (hunk 3 above)
Comment (prjemian) [reply user/Lambda.py]: done, needs discussion

**Comment by prjemian** [comment]
@sureshnaps , @qzhang234 : ready for you to review & discuss

**Comment by prjemian** [comment]
Thanks.  I'll open a new issue noting we have some questions to be resolved later.  Then I'll merge this PR.

**Comment by prjemian** [comment]
@sureshnaps , @qzhang234 : I'll merge this tomorrow afternoon.

**Comment by qzhang234** [comment]
@prjemian @sureshnaps Suresh and I added the scalars in the detector file and the temperature set point variables in the DM Workflow file. Is it Ok for us to merge?

**Comment by prjemian** [comment]
There are still a couple items in #61 to resolve but this PR can merge without those.  They'll take a little work to resolve and this PR is need to start on #36.

**Pull request merged**